Some updates I needed to get things up and running on target

ernstp opened the pull request:
I suspect you have been using an arm-toolchain in /usr/ on your host.
I don't have an arm toolchain there and got a number of errors.

Tested with Yocto 2.1

ernstp commented on recipes-dotnet/corefx/corefx_1.1.0.bb:
```diff
@@ -14,6 +14,9 @@ SRCREV = "release/${PV}"
 S = "${WORKDIR}/git"
 
+# Install the stripped binaries, the unstripped are smaller and causes Bus error
+INHIBIT_PACKAGE_STRIP = "1"
```
INSANE_SKIP_${PN} += "already-stripped"
would probably be better.

Tragetaschen commented on recipes-support/libunwind/libunwind_1.1.bbappend:
```diff
@@ -0,0 +1,9 @@
+SRC_URI_append += "\
+           file://libunwind.git-770152268807e460184b4152e23aba9c86601090.patch \
+           file://expand-arm-Validate-memory-before-access.patch \
+"
```
My setup doesn't have `libunwind_1.1.bb` anywhere. The only recipe I have is the one in `poky/meta/recipes-support/libunwind/libunwind_git.bb`. I checked the poky history and the `krogoth` branch is the last one having the `libunwind_1.1.bb` recipe.

Tragetaschen replied:
Ah, I just saw you are actually running on krogoth...

Is there a way to make this optional? When I use this as `libunwind_%.bbappend`, my libunwind build on morty breaks with `../../git/src/arm/Ginit.c:134:1: error: redefinition of 'validate_mem'`

ernstp replied:
I was thinking that a .bbappend without an original file would simply do nothing, but of course, it throws an error.
I wasn't sure if you wanted to target Morty or master... I know master has a later version which already includes these patches.

ernstp replied:
Yeah, we can do it optional, but since you don't have any branches targeting krogoth I think we can just leave it out.
My original idea was that it didn't do any harm and only applied if you had 1.1 in the first place, but it did do harm. :-)

ernstp replied:
It works if you only apply expand-arm-Validate-memory-before-access.patch ?

Tragetaschen commented:
Regarding the toolchain: No, it's the other way around :)

I tried very, very hard to make sure the toolchain from my native sysroot is used. Both upstream coreclr and corefx are currently built with a rather strict "Ubuntu-Linux-with-its-official-toolchain-in-/usr" mindset. The cmake infrastructure and `*.toolchain` files in addition to the `build.sh` script (and friends) are set up as if nothing around them had created a proper cross compile setup. For example when cross compiling for ARM, they hardcode the target as `armv7-linux-gnueabihf` (which is how it's called in the Ubuntu package), although Yocto calls it `arm-target-linux-gnueabi`

Yocto, `meta-llvm` and the cmake class on the other hand do set up things just right so that a simple call to cmake at the project root should suffice. I had tried that, but it failed. The main reason is that I couldn't find all the places in the cmake files where things are hard-coded as outlined above.

Tragetaschen commented:
Is it ok with you, if I start cherry-picking the easy stuff (like the curl dependency)?

Tragetaschen commented on recipes-dotnet/coreclr/coreclr/toolchain.patch:
```diff
@@ -20,16 +20,16 @@ index 84ae2a5..17fdf15 100644
  ## Specify the toolchain
 -set(TOOLCHAIN "arm-linux-gnueabihf")
-+set(TOOLCHAIN "arm-target-linux-gnueabi")
++set(TOOLCHAIN "arm-poky-linux-gnueabi")
```
Hmm, I had convinced myself that the compiler name is actually `arm-target-linux-gnueabi`, but it looks like the `target` and `poky` parts are the name of the distro (my distro is called `target` as in the company name of my employer).

That means this simple patch approach doesn't suffice :-/

ernstp replied:
Doh, I was also fooled by the generic name.
Ok, I'm not running original poky of course, I have customizations, but I didn't change to distribution name.

I guess the options are:
1) do_unpack_append () { sed -i s/arm-linux-gnueabihf/${YOCTO VARIABLES.../ }
2) template stuff on toolchain.patch.
Any preference?

Tragetaschen replied:
I'm fine with either approach

ernstp commented:
Btw, are you testing with Morty 2.2 or master?

And have you tried doing a lldb pull request to meta-clang?

Tragetaschen commented:
I'm on Morty. Not the latest, though, but what came out end of October.

Good point about the pull request...

Commit:
coreclr: Fix toolchain usage

The name of the yocto toolchain can vary, make sure we use the right
one. Clang uses binutils ld for linking, this
needs special handling to pick up.

Reviewed-by: Jens Nordenbro <jens.nordenbro@beijerelectronics.com>

Commit:
Add patch for libunwind

One patch is upstream in the Morty version and later, not the other
one.
https://github.com/dotnet/coreclr/blob/master/Documentation/building/linux-instructions.md#build-for-armlinux

Reviewed-by: Jens Nordenbro <jens.nordenbro@beijerelectronics.com>

Commit:
corefx: Avoid using the unstripped binaries

For some reason the stripped binaries (.so) are actually larger
than the unstripped (.so.dbg). Something is broken there, I get
a Bus error when running Hello world on target with them.
This was when testing with Yocto 2.1 Krogoth on an iMX6 platform.

Reviewed-by: Jens Nordenbro <jens.nordenbro@beijerelectronics.com>

Commit:
corefx: Fix toolchain usage

The name of the yocto toolchain can vary, make sure we use the right
one. Clang uses binutils ld for linking, this
needs special handling to pick up.

Reviewed-by: Jens Nordenbro <jens.nordenbro@beijerelectronics.com>

Commit:
Make corefx more similar to MS distribution

When you download dotnet core from
https://www.microsoft.com/net/download/linux
you get a distribution with some extra stuff.
This adds libuv support and downloaded Microsoft.CodeAnalysis dlls
to our package.
Still missing is libhostpolicy and libhostfxr, and the dotnet
executable.

Reviewed-by: Jens Nordenbro <jens.nordenbro@beijerelectronics.com>

ernstp commented:
Update series! Tried to put them in complexity order so you can fast-forward to a certain point if you want.

ernstp commented on recipes-dotnet/coreclr/coreclr/toolchain.patch:
```diff
@@ -1,8 +1,18 @@
+From 51d254e8993202a54c160ba52b412601090bd2cb Mon Sep 17 00:00:00 2001
```
Used devtool for editing, it added some extra cruft. Can remove that if you want to..

Tragetaschen commented on recipes-dotnet/coreclr/coreclr_1.1.0.bb:
```diff
@@ -16,14 +16,20 @@ SRCREV = "release/${PV}"
 S = "${WORKDIR}/git"
 
+do_fix_target_name() {
+	sed -i s/arm-linux-gnueabihf/${TARGET_SYS}/g cross/arm/toolchain.cmake
```
I've added an `${S}/` here before the cross. Otherwise sed complains it cannot find the file

ernstp replied:
Wierd that I didn't run in to any problems! Will fix.

Tragetaschen commented on recipes-dotnet/corefx/corefx_1.1.0.bb:
```diff
@@ -14,14 +14,23 @@ SRCREV = "release/${PV}"
+INSANE_SKIP_${PN} += "already-stripped dev-so"
+
+do_fix_target_name() {
+	sed -i s/arm-linux-gnueabihf/${TARGET_SYS}/g cross/arm/toolchain.cmake
```
see above

Tragetaschen commented on recipes-dotnet/corefx/corefx_1.1.0.bb:
```diff
@@ -34,19 +43,34 @@ do_install() {
+  Microsoft.CodeAnalysis.VisualBasic.dll \
+  Microsoft.CSharp.dll \
+"
 
```
This is Roslyn (not proprietary :-). It's likely this will become available through the `core-setup` and CLI recipes that are the next steps.

It's good to have them already, though.

JensNordenbro replied:
Correct me if I am wrong @ernstp, but I think we needed them to run aps.net core solutions using corerun. At this point we manage to run basic examples in mvc, websocket, static files using ssl. We have stripped out tooling-nugets that belongs to dotnet-cli capabilities however still *.CodeAnalysis.* was required.

Tragetaschen commented:
I have been successful in building everything with these commits.
It also runs on my platform.

After the `sed` fix, I'll be happy to merge.

ernstp commented:
I updated some of the commit messages to better describe what we're doing also!

Tragetaschen commented:
Thanks!
